File: Solution/【016】最接近的三数之和/最接近的三数之和.java
/*
注意此处没有之前类似三数之和这样的内层 while 提高效率，因为此处直接返回了
其他的和三数之和差不多
*/


class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if(nums == null || nums.length == 0)
            return 0;
        int min = Integer.MAX_VALUE;
        int result = 0;
        Arrays.sort(nums);
        for(int i = 0 ; i < nums.length ; i ++){
            int j = i + 1 ;
            int k = nums.length - 1;
            while( j < k ){
                if(Math.abs(nums[i] + nums[j] + nums[k] - target) < min){
                    min = Math.abs(nums[i] + nums[j] + nums[k] - target);
                    result = nums[i] + nums[j] + nums[k] ;
                }
                if(nums[i] + nums[j] + nums[k] == target){   //此处也是和 target的判断
                    return nums[i] + nums[j] + nums[k];
                }
                else if(nums[i] + nums[j] + nums[k] > target){   //有关target的判断
                    k = k - 1;
                }
                else{
                    j = j + 1;
                }
            }
        }
        return result;
    }
}

File: Solution/【016】最接近的三数之和/最接近的三数之和.cpp
/*
（1）一定要先排序   sort(nums.begin(), nums.end())
（2）计算 abs 一定要在最开始
*/



class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        if(nums.empty() || nums.size() == 0){
            return 0;
        }
        int result = 0;
        int temp = INT_MAX;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < nums.size(); i ++){
            int j = i + 1;
            int k = nums.size() - 1;
            while(j < k){
                if(abs(nums[i] + nums[j] + nums[k] - target) < temp){
                    temp = abs(nums[i] + nums[j] + nums[k] - target);
                    result = nums[i] + nums[j] + nums[k];
                }
                if(nums[i] + nums[j] + nums[k] == target){
                    return target;
                }
                else if(nums[i] + nums[j] + nums[k] > target){
                    k --;
                }
                else{
                    j ++;
                }
            }
        }
        return result;
    }
};